stdlib/public/core/BinaryFloatingPoint.swift

```diff
@@ -0,0 +1,677 @@
+    
+    let magnitude = source.magnitude
+    
+    let exponent = Int(magnitude._binaryLogarithm())
+    let exemplar = Self.greatestFiniteMagnitude
+    guard _fastPath(exponent <= exemplar.exponent) else {
+      return Source.isSigned && source < 0
+        ? (-.infinity, false)
+        : (.infinity, false)
+    }
+    let exponentBitPattern =
+      (1 as Self).exponentBitPattern /* (i.e., bias) */
+        + Self.RawExponent(exponent)
+    
+    let maxSignificandWidth = exemplar.significandWidth
+    let shift = maxSignificandWidth &- exponent
+    let significandBitPattern = shift >= 0
+      ? Self.RawSignificand(magnitude) << shift & exemplar.significandBitPattern
+      : Self.RawSignificand(
+        magnitude >> -shift & Source.Magnitude(exemplar.significandBitPattern))
+    
+    let value = Self(
+      sign: Source.isSigned && source < 0 ? .minus : .plus,
+      exponentBitPattern: exponentBitPattern,
+      significandBitPattern: significandBitPattern)
+    
+    if exponent &- magnitude.trailingZeroBitCount <= maxSignificandWidth {
+      return (value, true)
+    }
+    // We promise to round to the closest representation, and if two
+    // representable values are equally close, the value with more trailing
+    // zeros in its significand bit pattern. Therefore, we must take a look at
+    // the bits that we've just truncated.
+    let ulp = (1 as Source.Magnitude) << -shift
+    let truncatedBits = magnitude & (ulp - 1)
+    if truncatedBits < ulp / 2 {
+      return (value, false)
+    }
+    let rounded = Source.isSigned && source < 0 ? value.nextDown : value.nextUp
+    guard _fastPath(
+      truncatedBits != ulp / 2 ||
+        exponentBitPattern.trailingZeroBitCount <
+        rounded.exponentBitPattern.trailingZeroBitCount) else {
+          return (value, false)
+    }
+    return (rounded, false)
+  }
+  
+  /// Creates a new value, rounded to the closest possible representation.
+  ///
+  /// If two representable values are equally close, the result is the value
+  /// with more trailing zeros in its significand bit pattern.
+  ///
+  /// - Parameter value: The integer to convert to a floating-point value.
+  @inlinable // FIXME(sil-serialize-all)
+  public init<Source : BinaryInteger>(_ value: Source) {
+    self = Self._convert(from: value).value
+  }
+  
+  /// Creates a new value, if the given integer can be represented exactly.
+  ///
+  /// If the given integer cannot be represented exactly, the result is `nil`.
+  ///
+  /// - Parameter value: The integer to convert to a floating-point value.
+  @inlinable // FIXME(sil-serialize-all)
+  public init?<Source : BinaryInteger>(exactly value: Source) {
+    let (value_, exact) = Self._convert(from: value)
+    guard exact else { return nil }
+    self = value_
+  }
+  
+  @inlinable // FIXME(sil-serialize-all)
+  public // @testable
+  static func _convert<Source : BinaryFloatingPoint>(
+    from source: Source
+    ) -> (value: Self, exact: Bool) {
```

Comment:
I think moving this to a different file has done something with the indentation here. I know Xcode tends to do this when copying and pasting.

Reply:
Yup. Command-option-shift-v preserves formatting though, FYI.